Modernized ES modules in RecoTracker/MeasurementDet

Dr15Jones opened the pull request:
#### PR description:

- Added fillDescriptions
   - This required removing unused parameters from config files
- Used ESGetToken and consumes were needed
- Have ParameterSet only used by ESProducer and not helpers
- Combined header and source for ESProducer and removed modules.cc

#### PR validation:

Ran workflow 4.17 successfully after the change.

Dr15Jones committed:
Get parameters in Chi2ChargeMeasurementEstimatorESProducer c'tor

Read all the values from the edm::ParameterSet in the constructor.
This avoids the large memory needed to copy the edm::ParameterSet
and guarantees that all parameters are OK before starting the
data processing.

Dr15Jones committed:
Modernized MeasurementTrackerESProducer

- Used ESGetTokens and consumes
- Moved all handling of ParameterSet into MeasurementTrackerESProducer
- Added fillDescriptions

Dr15Jones committed:
Removed unused parameters for MeasurementTrackerESProducer

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27589/11008

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

RecoLocalTracker/SubCollectionProducers
RecoTracker/Configuration
RecoTracker/MeasurementDet


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@echabert, @pieterdavid, @makortel, @felicepantaleo, @yduhm, @GiacomoSguazzoni, @gbenelli, @rovere, @VinInn, @mschrode, @threus, @gpetruc, @ebrondol, @mmusich, @dgulhan, @alesaggio this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8ba462411de50b48aa4f40014e15aaf77c7c0117
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-438640/1598/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-438640/1598/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 2628546
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2628228
 - DQMHistoTests: Total skipped: 317
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

slava77 commented on RecoTracker/MeasurementDet/plugins/MeasurementTrackerESProducer.cc:
```diff
@@ -57,111 +167,84 @@ std::unique_ptr<MeasurementTracker> MeasurementTrackerESProducer::produce(const
+  desc.add<bool>("UsePixelROCQualityDB", true)->setComment("Use DB info at the ROC level");
+  desc.addUntracked<bool>("DebugPixelROCQualityDB", false)->setComment("dump out info om module status ");
+
+  descriptions.addDefault(desc);
```
what is the reason there is no named description to replace MeasurementTrackerESProducer_cfi ?

Dr15Jones replied:
As the Core is adding `consumes` to all ESProducers, we've tried to minimize the possible impact of our changes (although sometimes we just can't help ourselves and feel compelled to re-write more code then necessary :) ). So I didn't want to touch the cfi.py files actually loaded in production. Plus MeasurementTrackerESProducer_cfi.py uses an `Era` which can not be done by the `fillDescriptions`.

slava77 replied:
I worry that this is opening up a problem of large and complex duplication of the configuration with different implementation, which is bound to quickly diverge and become confusing.

Dr15Jones replied:
I can remove all the default values if you prefer.

perrotta replied:
... and confuse HLT parsing as well.

One could replace courrent _cfi with a _cff one, which imports the auto-generated one and can still implemet the Era's. If so, one should also propagate whenever there is an inclusion of MeasurementTrackerESProducer_cfi.py  with the new MeasurementTrackerESProducer_cff.py

perrotta replied:
Or even let the fillDescriptions define a MeasurementTrackerESProducerDefault_cfi (or similar)

Dr15Jones replied:
I have been considering having the `addDefault` actually allow generation of a `classtype_default.py` file which would declare a variable with name `_classtype_default`. The `_` guarantees that an `import *` statement or doing `process.load` would not pick it up accidentally but it still could be used for cloning.

slava77 replied:
thinking of some other ways we were getting to the era-modified cases, this pattern seems reasonable to me (a generated cfi has a clearly distinct, not intuitively usable general name, the manually created cfi or cff would load it and clone to the usable name and also have the modifiers applied as necessary)

Dr15Jones replied:
So what do you want me to do?
1. leave as is
1. remove all default values in fillDescriptions except "Phase2StripCPE" which was previously using the `existsAs` trick to create a default value for a tracked parameter (which we do NOT want people to do)
1. have the fillDescriptions write a `MeasurementTrackerESProducerDefault_cfi.py` file and then change `MeasurementTrackerESProducer_cfi.py` to load that file in such a way as to not create 2 modules

slava77 replied:
I do not want to have \~30 parameters (with somewhat complicated nested structure) defined in two places with some expectations of identity.
[the limit where it still seems acceptable to me is around O(1) parameter]

The proposal to generate ```classtype_default.py``` (perhaps still a \*cfi.py) looks good to me (perhaps I was not clear in https://github.com/cms-sw/cmssw/pull/27589#discussion_r307390394).
Please proceed.
Thank you.

Dr15Jones replied:
What I had proposed for `classtype_default.py` was a possible future extension to the parameter set system and would not be part of these pull requests. The 'option 3' (make a `MeasurementTrackerESProducerDefault_cfi.py` file using fillDescriptions) approximates that behavior based on what is possible now.

I will do 'option 3'.

slava77 commented:
@Dr15Jones 
please clarify if you are planning to update this PR (in line with https://github.com/cms-sw/cmssw/pull/27589#discussion_r307349832 ).
Thank you.

perrotta commented:
(Similar issues for #27594 and #27596)

Dr15Jones commented:
Sorry, I was waiting for @slava77 and @perrotta to decide how you wanted to handle the fillDescriptions as it is code you'd have to maintain. From my side, doing `addDefault` is the least intrusive way to improve the code.

Dr15Jones committed:
fillDescriptions now creates a cfi with the default values

The original MeasurementTrackerESProducer_cfi.py now imports the
new default cfi and uses it as the starting values.
This was the agreement with the reconstruction L2s.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27589/11216

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #27589 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e65424cba99b7dc1d8c359d475d15c8a47f6d81a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-438640/1752/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-438640/1752/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-438640/22034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D46_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D46+RecoFullGlobal_2026D46+HARVESTFullGlobal_2026D46

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 2484130
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2483811
 - DQMHistoTests: Total skipped: 317
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 132 log files, 14 edm output root files, 32 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/27589	e65424c
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

MeasurementTracker in  RecoTracker/MeasurementDet/python/MeasurementTrackerESProducer_cfi.py is now appropriately derived from the fillDescriptions

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

kpedro88 commented:
+1

makortel commented:
part of #26748
